File: packages/backend/src/repo/dynamo.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
  QueryCommand,
} from '@aws-sdk/lib-dynamodb';
import type {
  Booking,
  ChatSession,
  MatchResult,
  ServiceCategory,
  ServiceRequest,
  UserPreferences,
  UserProfile,
  Vendor,
} from '../domain/types';
import { config } from '../lib/config';
import { mergePrefs } from './memory';
import type { Repo } from './types';
import { vendorCovers } from './coverage';

/**
 * DynamoDB single-table design
 *
 * | 實體      | PK                  | SK        | GSI1PK        | GSI1SK              |
 * |-----------|---------------------|-----------|---------------|---------------------|
 * | 會員      | USER#<accountId>    | PROFILE   | -             | -                   |
 * | 廠商      | VENDOR#<vendorId>   | META      | VENDOR        | <vendorId>          |
 * | 服務單    | REQ#<requestId>     | META      | USER#<acc>    | REQ#<createdAt>     |
 * | 媒合結果  | REQ#<requestId>     | MATCH     | -             | -                   |
 * | 預約單    | BOOKING#<orderNo>   | META      | USER#<acc>    | BOOKING#<createdAt> |
 * | 對話      | SESSION#<sessionId> | META      | -             | -                   |
 */
const GSI1 = 'GSI1';

const doc = DynamoDBDocumentClient.from(new DynamoDBClient({ region: config.region }), {
  marshallOptions: { removeUndefinedValues: true },
});

interface Item {
  PK: string;
  SK: string;
  GSI1PK?: string;
  GSI1SK?: string;
  data: unknown;
}

export class DynamoRepo implements Repo {
  private table = config.tableName;

  private async get<T>(PK: string, SK: string): Promise<T | undefined> {
    const res = await doc.send(new GetCommand({ TableName: this.table, Key: { PK, SK } }));
    return (res.Item as Item | undefined)?.data as T | undefined;
  }

  private async put(item: Item): Promise<void> {
    await doc.send(new PutCommand({ TableName: this.table, Item: item }));
  }

  private async queryGsi1<T>(pk: string, skPrefix?: string): Promise<T[]> {
    const res = await doc.send(
      new QueryCommand({
        TableName: this.table,
        IndexName: GSI1,
        KeyConditionExpression: skPrefix
          ? 'GSI1PK = :pk AND begins_with(GSI1SK, :sk)'
          : 'GSI1PK = :pk',
        ExpressionAttributeValues: skPrefix ? { ':pk': pk, ':sk': skPrefix } : { ':pk': pk },
      }),
    );
    return (res.Items ?? []).map((i) => (i as Item).data as T);
  }

  // ---- 會員 ----
  getUser(id: string) {
    return this.get<UserProfile>(`USER#${id}`, 'PROFILE');
  }

  putUser(user: UserProfile) {
    return this.put({ PK: `USER#${user.inbrAccountId}`, SK: 'PROFILE', data: user });
  }

  async mergePreferences(id: string, patch: UserPreferences): Promise<UserPreferences> {
    const user = await this.getUser(id);
    if (!user) throw new Error(`user not found: ${id}`);
    user.preferences = mergePrefs(user.preferences, patch);
    await this.putUser(user);
    return user.preferences;
  }

  // ---- 廠商 ----
  async listVendors(filter?: {
    category?: ServiceCategory;
    countyCode?: string;
    districtCode?: string;
  }) {
    let list = await this.queryGsi1<Vendor>('VENDOR');
    if (filter?.category) list = list.filter((v) => v.categories.includes(filter.category!));
    if (filter?.countyCode) {
      list = list.filter((v) => vendorCovers(v, filter.countyCode!, filter.districtCode));
    }
    return list;
  }

  getVendor(vendorId: string) {
    return this.get<Vendor>(`VENDOR#${vendorId}`, 'META');
  }

  putVendor(vendor: Vendor) {
    return this.put({
      PK: `VENDOR#${vendor.vendorId}`,
      SK: 'META',
      GSI1PK: 'VENDOR',
      GSI1SK: vendor.vendorId,
      data: vendor,
    });
  }

  // ---- 服務單 ----
  getRequest(requestId: string) {
    return this.get<ServiceRequest>(`REQ#${requestId}`, 'META');
  }

  putRequest(req: ServiceRequest) {
    return this.put({
      PK: `REQ#${req.requestId}`,
      SK: 'META',
      GSI1PK: `USER#${req.inbrAccountId}`,
      GSI1SK: `REQ#${req.createdAt}`,
      data: req,
    });
  }

  listRequestsByUser(id: string) {
    return this.queryGsi1<ServiceRequest>(`USER#${id}`, 'REQ#');
  }

  // ---- 媒合結果 ----
  putMatch(match: MatchResult) {
    return this.put({ PK: `REQ#${match.requestId}`, SK: 'MATCH', data: match });
  }

  getMatch(requestId: string) {
    return this.get<MatchResult>(`REQ#${requestId}`, 'MATCH');
  }

  // ---- 預約單 ----
  putBooking(booking: Booking) {
    return this.put({
      PK: `BOOKING#${booking.orderNo}`,
      SK: 'META',
      GSI1PK: `USER#${booking.inbrAccountId}`,
      GSI1SK: `BOOKING#${booking.createdAt}`,
      data: booking,
    });
  }

  getBooking(orderNo: string) {
    return this.get<Booking>(`BOOKING#${orderNo}`, 'META');
  }

  listBookingsByUser(id: string) {
    return this.queryGsi1<Booking>(`USER#${id}`, 'BOOKING#');
  }

  // ---- 對話 ----
  getSession(sessionId: string) {
    return this.get<ChatSession>(`SESSION#${sessionId}`, 'META');
  }

  putSession(session: ChatSession) {
    return this.put({ PK: `SESSION#${session.sessionId}`, SK: 'META', data: session });
  }
}

File: packages/api/op_agent/repo/dynamo.py
"""DynamoDB single-table 實作。

| 實體      | PK                  | SK      | GSI1PK      | GSI1SK              |
|-----------|---------------------|---------|-------------|---------------------|
| 會員      | USER#<accountId>    | PROFILE | -           | -                   |
| 廠商      | VENDOR#<vendorId>   | META    | VENDOR      | <vendorId>          |
| 服務單    | REQ#<requestId>     | META    | USER#<acc>  | REQ#<createdAt>     |
| 媒合結果  | REQ#<requestId>     | MATCH   | -           | -                   |
| 預約單    | BOOKING#<orderNo>   | META    | USER#<acc>  | BOOKING#<createdAt> |
| 對話      | SESSION#<sessionId> | META    | -           | -                   |

DynamoDB 不吃 float，所以寫入前把 float 轉 Decimal、讀出時轉回去。
"""

from __future__ import annotations

from decimal import Decimal
from typing import Any

import boto3
from boto3.dynamodb.conditions import Key

from ..config import config
from ..domain import (
    Booking,
    ChatSession,
    MatchResult,
    ServiceRequest,
    UserPreferences,
    UserProfile,
    Vendor,
)
from .base import merge_prefs, vendor_covers

GSI1 = "GSI1"


def _to_dynamo(value: Any) -> Any:
    """float -> Decimal，並丟掉 None（DynamoDB 可存 null 但這裡簡化資料）。"""
    if isinstance(value, float):
        return Decimal(str(value))
    if isinstance(value, dict):
        return {k: _to_dynamo(v) for k, v in value.items() if v is not None}
    if isinstance(value, list):
        return [_to_dynamo(v) for v in value]
    return value


def _from_dynamo(value: Any) -> Any:
    """Decimal -> int/float。"""
    if isinstance(value, Decimal):
        return int(value) if value == value.to_integral_value() else float(value)
    if isinstance(value, dict):
        return {k: _from_dynamo(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_from_dynamo(v) for v in value]
    return value


class DynamoRepo:
    def __init__(self) -> None:
        self._table = boto3.resource("dynamodb", region_name=config.region).Table(
            config.table_name
        )

    # ---- 低階讀寫 ----
    def _get(self, pk: str, sk: str) -> Any | None:
        res = self._table.get_item(Key={"PK": pk, "SK": sk})
        item = res.get("Item")
        return _from_dynamo(item["data"]) if item else None

    def _put(
        self, pk: str, sk: str, data: Any, gsi1pk: str | None = None, gsi1sk: str | None = None
    ) -> None:
        item: dict[str, Any] = {"PK": pk, "SK": sk, "data": _to_dynamo(data)}
        if gsi1pk and gsi1sk:
            item["GSI1PK"] = gsi1pk
            item["GSI1SK"] = gsi1sk
        self._table.put_item(Item=item)

    def _query_gsi1(self, pk: str, sk_prefix: str | None = None) -> list[Any]:
        cond = Key("GSI1PK").eq(pk)
        if sk_prefix:
            cond = cond & Key("GSI1SK").begins_with(sk_prefix)
        items: list[Any] = []
        kwargs: dict[str, Any] = {"IndexName": GSI1, "KeyConditionExpression": cond}
        while True:
            res = self._table.query(**kwargs)
            items.extend(_from_dynamo(i["data"]) for i in res.get("Items", []))
            key = res.get("LastEvaluatedKey")
            if not key:
                return items
            kwargs["ExclusiveStartKey"] = key

    # ---- 會員 ----
    def get_user(self, inbr_account_id: str) -> UserProfile | None:
        return self._get(f"USER#{inbr_account_id}", "PROFILE")

    def put_user(self, user: UserProfile) -> None:
        self._put(f"USER#{user['inbrAccountId']}", "PROFILE", user)

    def merge_preferences(self, inbr_account_id: str, patch: UserPreferences) -> UserPreferences:
        user = self.get_user(inbr_account_id)
        if user is None:
            raise KeyError(f"user not found: {inbr_account_id}")
        user["preferences"] = merge_prefs(user.get("preferences", {}), patch)
        self.put_user(user)
        return user["preferences"]

    # ---- 廠商 ----
    def list_vendors(
        self,
        *,
        category: str | None = None,
        county_code: str | None = None,
        district_code: str | None = None,
    ) -> list[Vendor]:
        items: list[Vendor] = self._query_gsi1("VENDOR")
        if category:
            items = [v for v in items if category in v.get("categories", [])]
        if county_code:
            items = [v for v in items if vendor_covers(v, county_code, district_code)]
        return items

    def get_vendor(self, vendor_id: str) -> Vendor | None:
        return self._get(f"VENDOR#{vendor_id}", "META")

    def put_vendor(self, vendor: Vendor) -> None:
        self._put(f"VENDOR#{vendor['vendorId']}", "META", vendor, "VENDOR", vendor["vendorId"])

    # ---- 服務單 ----
    def get_request(self, request_id: str) -> ServiceRequest | None:
        return self._get(f"REQ#{request_id}", "META")

    def put_request(self, req: ServiceRequest) -> None:
        self._put(
            f"REQ#{req['requestId']}",
            "META",
            req,
            f"USER#{req['inbrAccountId']}",
            f"REQ#{req['createdAt']}",
        )

    def list_requests_by_user(self, inbr_account_id: str) -> list[ServiceRequest]:
        return self._query_gsi1(f"USER#{inbr_account_id}", "REQ#")

    # ---- 媒合結果 ----
    def put_match(self, match: MatchResult) -> None:
        self._put(f"REQ#{match['requestId']}", "MATCH", match)

    def get_match(self, request_id: str) -> MatchResult | None:
        return self._get(f"REQ#{request_id}", "MATCH")

    # ---- 預約單 ----
    def put_booking(self, booking: Booking) -> None:
        self._put(
            f"BOOKING#{booking['orderNo']}",
            "META",
            booking,
            f"USER#{booking['inbrAccountId']}",
            f"BOOKING#{booking['createdAt']}",
        )

    def get_booking(self, order_no: str) -> Booking | None:
        return self._get(f"BOOKING#{order_no}", "META")

    def list_bookings_by_user(self, inbr_account_id: str) -> list[Booking]:
        return self._query_gsi1(f"USER#{inbr_account_id}", "BOOKING#")

    # ---- 對話 ----
    def get_session(self, session_id: str) -> ChatSession | None:
        return self._get(f"SESSION#{session_id}", "META")

    def put_session(self, session: ChatSession) -> None:
        self._put(f"SESSION#{session['sessionId']}", "META", session)
